### ISSUE TITLE
cairo/1.15.14@bincrafters/stable for Raspberry Pi armv7hf

### ISSUE THREAD

**Issue by izzyreal** [bincrafters/community#706]
Anyone been able to do a conan install of that package against armv7hf? I'm trying in [https://hub.docker.com/r/conanio/gcc7-armv7hf/](https://hub.docker.com/r/conanio/gcc7-armv7hf/) and I'm getting various redefinition errors, the first one being
```
cairo-xlib-xrender-private.h:106:16: error: redefinition of ‘struct _XLinearGradient’
 typedef struct _XLinearGradient {
                ^~~~~~~~~~~~~~~~
In file included from cairo-xlib-xrender.h:45:0,
                 from cairo-xlib-xrender-private.h:53,
                 from cairo-xlib-private.h:40,
                 from cairo-xlib-display.c:40:
/usr/include/X11/extensions/Xrender.h:186:16: note: originally defined here
 typedef struct _XLinearGradient {
                ^~~~~~~~~~~~~~~~
In file included from cairo-xlib-private.h:40:0,
                 from cairo-xlib-display.c:40:
cairo-xlib-xrender-private.h:109:3: error: conflicting types for ‘XLinearGradient’
 } XLinearGradient;
   ^~~~~~~~~~~~~~~
In file included from cairo-xlib-xrender.h:45:0,
                 from cairo-xlib-xrender-private.h:53,
                 from cairo-xlib-private.h:40,
                 from cairo-xlib-display.c:40:
/usr/include/X11/extensions/Xrender.h:189:3: note: previous declaration of ‘XLinearGradient’ was here
 } XLinearGradient;
   ^~~~~~~~~~~~~~~
```

**Comment by izzyreal** [comment]
Hi @uilianries, to reproduce the error
* Download docker image [https://hub.docker.com/r/conanio/gcc7-armv7hf/](https://hub.docker.com/r/conanio/gcc7-armv7hf/)
* Run `docker run -it --rm conanio/conangcc7-armv7hf /bin/bash`
* Run `conan remote add bincrafters https://api.bintray.com/conan/bincrafters/public-conan`
* Run `conan install cairo/1.15.14@bincrafters/stable --build cairo`

**Comment by izzyreal** [comment]
If I change cairo's conanfile.py line 103 to `configure_args = ['--disable-ft', '--disable-xlib']`, so adding the disable-xlib flag, cairo builds fine.

**Comment by uilianries** [comment]
Thanks for the details @izzyreal !

I'll provide a fix later, thanks again!

**Comment by uilianries** [comment]
That's weird ... I can't reproduce your error. What I've done:

    docker run --rm -t conanio/gcc7-armv7hf /bin/bash -c "conan remote add bincrafters https://api.bintray.com/conan/bincrafters/public-conan && conan install cairo/1.15.14@bincrafters/stable --build missing"

The result is all build process with success.

Did you update your docker image?

My current image is:

	REPOSITORY                                   TAG                    IMAGE ID            CREATED             SIZE
    conanio/gcc7-armv7hf                         latest                 46b6e0532c82        32 hours ago        1.37GB

**Comment by izzyreal** [comment]
Hm 32 hours ago... I may not have that latest one then, I will try again and report here. Thx!

**Comment by izzyreal** [comment]
Ok confirmed to build fine indeed :) I must have fux03r3d something in my previous image or indeed the image was updated.

**Comment by uilianries** [comment]
Thanks for replying. I gonna this issue, feel free to re-open.

**Comment by izzyreal** [comment]
@uilianries Not sure how to re-open. The issue returned when first installing SDL2.

**Comment by uilianries** [comment]
Please, post here which commands did you run and which docker image.

**Comment by izzyreal** [comment]
* Download latest docker image [https://hub.docker.com/r/conanio/gcc7-armv7hf/](https://hub.docker.com/r/conanio/gcc7-armv7hf/)
* Run `docker run -it --rm conanio/conangcc7-armv7hf /bin/bash`
* Run `conan remote add bincrafters https://api.bintray.com/conan/bincrafters/public-conan`
* Run `conan install sdl2/2.0.9@bincrafters/stable --build missing` and it will fail
* Edit sdl2 conanfile.py by adding
```
                elif self.settings.arch == "armv7hf":
                    arch_suffix = ':armhf'
```
* Run `conan install sdl2/2.0.9@bincrafters/stable --build missing` again and it will fail
* Edit Xlib.h as per [https://github.com/anyc/steam-overlay/issues/61#issuecomment-20571907](https://github.com/anyc/steam-overlay/issues/61#issuecomment-20571907)
* Run `conan install sdl2/2.0.9@bincrafters/stable --build missing` again and it will succeed
* Run `conan install cairo/1.15.14@bincrafters/stable --build cairo` and it will fail:
```
cairo-xlib-xrender-private.h:106:16: error: redefinition of ‘struct _XLinearGradient’
 typedef struct _XLinearGradient {
                ^~~~~~~~~~~~~~~~
In file included from cairo-xlib-xrender.h:45:0,
                 from cairo-xlib-xrender-private.h:53,
                 from cairo-xlib-private.h:40,
                 from cairo-xlib-display.c:40:
/usr/include/X11/extensions/Xrender.h:186:16: note: originally defined here
 typedef struct _XLinearGradient {
                ^~~~~~~~~~~~~~~~
In file included from cairo-xlib-private.h:40:0,
                 from cairo-xlib-display.c:40:
cairo-xlib-xrender-private.h:109:3: error: conflicting types for ‘XLinearGradient’
 } XLinearGradient;
   ^~~~~~~~~~~~~~~
In file included from cairo-xlib-xrender.h:45:0,
                 from cairo-xlib-xrender-private.h:53,
                 from cairo-xlib-private.h:40,
                 from cairo-xlib-display.c:40:
/usr/include/X11/extensions/Xrender.h:189:3: note: previous declaration of ‘XLinearGradient’ was here
 } XLinearGradient;
   ^~~~~~~~~~~~~~~
```

**Comment by Croydon** [comment]
This is also a combination of different problems.

Let's fix first: #706, then #714 then have a look at this one again

**Comment by Croydon** [comment]
This recipe is now in the [Conan Center Index](https://github.com/conan-io/conan-center-index/).

Please migrate to the recipe version there. 
If this is still an issue, please create a new issue in the [Conan Center Index issue tracker](https://github.com/conan-io/conan-center-index/issues).

Thanks!
